![imagenes](logo.png)

# SIUBA

El producto interno bruto (PIB) es una medida monetaria del valor de mercado de todos los bienes y servicios finales producidos y vendidos (no revendidos) en un período de tiempo específico por países.

El dataframe ``gapminder`` presenta información del PIB de varios países desde el año 1952 hasta el 2007 por periodos de 5 años.

Su informacion viene dada por las columnas:

* ``country``: país
* ``continent``: continente
* ``year``: año
* ``lifeExp``: esperanza de vida
* ``population``: población
* ``gdpPercap`` PIB per cápita

In [2]:
## Mandamos a llamar nuestras bibliotecas principales

import pandas as pd
import numpy as np
from siuba import *
from siuba.dply.vector import *

In [4]:
## Leemos el archivo gapminder.csv
gapminder = pd.read_csv("C:\MAMP\Git\CDD_Py_2022\data\gapminder.csv")

## Muestra las primeras filas de gapminder

In [5]:
gapminder >> head()  #gapminder.head()

country continent  year  lifeExp  population   gdpPercap
0  Afghanistan      Asia  1952   28.801     8425333  779.445314
1  Afghanistan      Asia  1957   30.332     9240934  820.853030
2  Afghanistan      Asia  1962   31.997    10267083  853.100710
3  Afghanistan      Asia  1967   34.020    11537966  836.197138
4  Afghanistan      Asia  1972   36.088    13079460  739.981106

## Piensa antes de trabajar los datos

Una variable, o columna, de tipo categórica es aquella que indica si un renglón pertenece a alguna clase. Por ejemplo, en 

nombre|sexo|estatura
:-:|:-:|:-:
Luis|masc|1.70|
María|feme|1.65
Marco|masc|1.83
Helena|feme|1.61
Luisa|feme|1.67

la columna ``sexo`` nos indica si el renglón corresponde a la clase ``masc`` o a la clase ``feme``. A los diferentes valores que puede tomar una columna que utilizamos para clasificar se les llama *niveles* o *factores*.

Si tienes la necesidad de almacenar un pequeño fragmento de sus datos, detente y pregúntate **¿Quiero crear mini conjuntos de datos para cada nivel de alguna clase (o combinación única de varias columnas categóricas) para calcular o graficar algo?**

En **caso afirmativo**, usa técnicas de agregación de datos adecuadas. **No dividas los datos en subconjuntos**. O solo divide los datos en subconjuntos como una medida temporal mientras desarrollas tu código para computar o visualizar estos subconjuntos de datos.

Si **NO**, entonces tal vez realmente necesitas almacenar una copia de un subconjunto de los datos.

Toma en cuenta que las copias y extractos de tus datos saturan tu espacio de trabajo, invitan a cometer errores y generan confusión general. 

La realidad también puede estar en algún punto intermedio. Encontrarás que los flujos de trabajo que se presentan a continuación pueden ayudarte a lograr tus objetivos con una creación mínima de objetos intermedios temporales.

## filter() para crear subconjuntos de datos por filas.

``filter()`` toma expresiones lógicas y devuelve las filas para las que todos esas expresiones son verdaderas.

**Objetivos:**

* Encuentra las observaciones donde la esperanza de vida sea inferior a 29
* Encuentra las observaciones de Ruanda después del año 1979 (Ruanda en inglés es Rwanda)
* Encuentra las observaciones tanto de Ruanda como de Afganistán (Afganistán en inglés es Afghanistan)

In [8]:
gapminder >> filter(_.lifeExp < 29)  
# gapminder[gapminder["lifeExp"] < 29]

country continent  year  lifeExp  population   gdpPercap
0     Afghanistan      Asia  1952   28.801     8425333  779.445314
1292       Rwanda    Africa  1992   23.599     7290203  737.068595

In [12]:
gapminder >> filter(_.country == "Rwanda", _.year > 1979) 
# gapminder[(gapminder["country"] == "Rwanda") & (gapminder["year"] > 1979)]

country continent  year  lifeExp  population   gdpPercap
1290  Rwanda    Africa  1982   46.218     5507565  881.570647
1291  Rwanda    Africa  1987   44.020     6349365  847.991217
1292  Rwanda    Africa  1992   23.599     7290203  737.068595
1293  Rwanda    Africa  1997   36.087     7212583  589.944505
1294  Rwanda    Africa  2002   43.413     7852401  785.653765
1295  Rwanda    Africa  2007   46.242     8860588  863.088464

In [13]:
gapminder >> filter((_.country == "Rwanda") | (_.country == "Afghanistan"))
# gapminder[(gapminder["country"] == "Rwanda") | (gapminder["country"] == "Afghanistan")]

country continent  year  lifeExp  population   gdpPercap
0     Afghanistan      Asia  1952   28.801     8425333  779.445314
1     Afghanistan      Asia  1957   30.332     9240934  820.853030
2     Afghanistan      Asia  1962   31.997    10267083  853.100710
3     Afghanistan      Asia  1967   34.020    11537966  836.197138
4     Afghanistan      Asia  1972   36.088    13079460  739.981106
5     Afghanistan      Asia  1977   38.438    14880372  786.113360
6     Afghanistan      Asia  1982   39.854    12881816  978.011439
7     Afghanistan      Asia  1987   40.822    13867957  852.395945
8     Afghanistan      Asia  1992   41.674    16317921  649.341395
9     Afghanistan      Asia  1997   41.763    22227415  635.341351
10    Afghanistan      Asia  2002   42.129    25268405  726.734055
11    Afghanistan      Asia  2007   43.828    31889923  974.580338
1284       Rwanda    Africa  1952   40.000     2534927  493.323875
1285       Rwanda    Africa  1957   41.500     2822082  540.289398
1286       Rwanda    Africa  1962   43.000     3051242  597.473073
1287       Rwanda    Africa  1967   44.100     3451079  510.963714
1288       Rwanda    Africa  1972   44.600     3992121  590.580664
1289       Rwanda    Africa  1977   45.000     4657072  670.080601
1290       Rwanda    Africa  1982   46.218     5507565  881.570647
1291       Rwanda    Africa  1987   44.020     6349365  847.991217
1292       Rwanda    Africa  1992   23.599     7290203  737.068595
1293       Rwanda    Africa  1997   36.087     7212583  589.944505
1294       Rwanda    Africa  2002   43.413     7852401  785.653765
1295       Rwanda    Africa  2007   46.242     8860588  863.088464

### El método ``.isin()`` de pandas

Alternativamente, podemos emplear el método ``.isin()`` de pandas:

``columna.isin(<<lista>>)``


In [15]:
gapminder >> filter(_.country.isin(["Rwanda","Afghanistan"]))
# gapminder[gapminder["country"].isin(["Rwanda","Afghanistan"])]

country continent  year  lifeExp  population   gdpPercap
0     Afghanistan      Asia  1952   28.801     8425333  779.445314
1     Afghanistan      Asia  1957   30.332     9240934  820.853030
2     Afghanistan      Asia  1962   31.997    10267083  853.100710
3     Afghanistan      Asia  1967   34.020    11537966  836.197138
4     Afghanistan      Asia  1972   36.088    13079460  739.981106
5     Afghanistan      Asia  1977   38.438    14880372  786.113360
6     Afghanistan      Asia  1982   39.854    12881816  978.011439
7     Afghanistan      Asia  1987   40.822    13867957  852.395945
8     Afghanistan      Asia  1992   41.674    16317921  649.341395
9     Afghanistan      Asia  1997   41.763    22227415  635.341351
10    Afghanistan      Asia  2002   42.129    25268405  726.734055
11    Afghanistan      Asia  2007   43.828    31889923  974.580338
1284       Rwanda    Africa  1952   40.000     2534927  493.323875
1285       Rwanda    Africa  1957   41.500     2822082  540.289398
1286       Rwanda    Africa  1962   43.000     3051242  597.473073
1287       Rwanda    Africa  1967   44.100     3451079  510.963714
1288       Rwanda    Africa  1972   44.600     3992121  590.580664
1289       Rwanda    Africa  1977   45.000     4657072  670.080601
1290       Rwanda    Africa  1982   46.218     5507565  881.570647
1291       Rwanda    Africa  1987   44.020     6349365  847.991217
1292       Rwanda    Africa  1992   23.599     7290203  737.068595
1293       Rwanda    Africa  1997   36.087     7212583  589.944505
1294       Rwanda    Africa  2002   43.413     7852401  785.653765
1295       Rwanda    Africa  2007   46.242     8860588  863.088464

In [18]:
# Imprimo la cantidad de filas y columnas que tiene el dataframe
(gapminder >> filter(_.country.isin(["Rwanda","Afghanistan"]))).shape

(24, 6)

In [27]:
# Imprimo los paises que inician con alguna cadena de texto en este caso Guinea
gapminder >> filter(_.country.str.startswith(("Guinea","Guate")))

country continent  year  lifeExp  population    gdpPercap
600      Guatemala  Americas  1952   42.023     3146381  2428.237769
601      Guatemala  Americas  1957   44.142     3640876  2617.155967
602      Guatemala  Americas  1962   46.954     4208858  2750.364446
603      Guatemala  Americas  1967   50.016     4690773  3242.531147
604      Guatemala  Americas  1972   53.738     5149581  4031.408271
605      Guatemala  Americas  1977   56.029     5703430  4879.992748
606      Guatemala  Americas  1982   58.137     6395630  4820.494790
607      Guatemala  Americas  1987   60.782     7326406  4246.485974
608      Guatemala  Americas  1992   63.373     8486949  4439.450840
609      Guatemala  Americas  1997   66.322     9803875  4684.313807
610      Guatemala  Americas  2002   68.978    11178650  4858.347495
611      Guatemala  Americas  2007   70.259    12572928  5186.050003
612         Guinea    Africa  1952   33.609     2664249   510.196492
613         Guinea    Africa  1957   34.558     2876726   576.267025
614         Guinea    Africa  1962   35.753     3140003   686.373674
615         Guinea    Africa  1967   37.197     3451418   708.759541
616         Guinea    Africa  1972   38.842     3811387   741.666231
617         Guinea    Africa  1977   40.762     4227026   874.685864
618         Guinea    Africa  1982   42.891     4710497   857.250358
619         Guinea    Africa  1987   45.552     5650262   805.572472
620         Guinea    Africa  1992   48.576     6990574   794.348438
621         Guinea    Africa  1997   51.455     8048834   869.449767
622         Guinea    Africa  2002   53.676     8807818   945.583584
623         Guinea    Africa  2007   56.007     9947814   942.654211
624  Guinea-Bissau    Africa  1952   32.500      580653   299.850319
625  Guinea-Bissau    Africa  1957   33.489      601095   431.790457
626  Guinea-Bissau    Africa  1962   34.488      627820   522.034373
627  Guinea-Bissau    Africa  1967   35.492      601287   715.580640
628  Guinea-Bissau    Africa  1972   36.486      625361   820.224588
629  Guinea-Bissau    Africa  1977   37.465      745228   764.725963
630  Guinea-Bissau    Africa  1982   39.327      825987   838.123967
631  Guinea-Bissau    Africa  1987   41.245      927524   736.415392
632  Guinea-Bissau    Africa  1992   43.266     1050938   745.539871
633  Guinea-Bissau    Africa  1997   44.873     1193708   796.664468
634  Guinea-Bissau    Africa  2002   45.504     1332459   575.704718
635  Guinea-Bissau    Africa  2007   46.388     1472041   579.231743

In [32]:
# Imprimo la cantidad de filas y columnas si en una columna puntual que contenga una frase
gapminder >> filter(_.country.str.contains("Guinea|Mexico"))

country continent  year  lifeExp  population     gdpPercap
480  Equatorial Guinea    Africa  1952   34.482      216964    375.643123
481  Equatorial Guinea    Africa  1957   35.983      232922    426.096408
482  Equatorial Guinea    Africa  1962   37.485      249220    582.841971
483  Equatorial Guinea    Africa  1967   38.987      259864    915.596003
484  Equatorial Guinea    Africa  1972   40.516      277603    672.412257
485  Equatorial Guinea    Africa  1977   42.024      192675    958.566812
486  Equatorial Guinea    Africa  1982   43.662      285483    927.825343
487  Equatorial Guinea    Africa  1987   45.664      341244    966.896815
488  Equatorial Guinea    Africa  1992   47.545      387838   1132.055034
489  Equatorial Guinea    Africa  1997   48.245      439971   2814.480755
490  Equatorial Guinea    Africa  2002   49.348      495627   7703.495900
491  Equatorial Guinea    Africa  2007   51.579      551201  12154.089750
612             Guinea    Africa  1952   33.609     2664249    510.196492
613             Guinea    Africa  1957   34.558     2876726    576.267025
614             Guinea    Africa  1962   35.753     3140003    686.373674
615             Guinea    Africa  1967   37.197     3451418    708.759541
616             Guinea    Africa  1972   38.842     3811387    741.666231
617             Guinea    Africa  1977   40.762     4227026    874.685864
618             Guinea    Africa  1982   42.891     4710497    857.250358
619             Guinea    Africa  1987   45.552     5650262    805.572472
620             Guinea    Africa  1992   48.576     6990574    794.348438
621             Guinea    Africa  1997   51.455     8048834    869.449767
622             Guinea    Africa  2002   53.676     8807818    945.583584
623             Guinea    Africa  2007   56.007     9947814    942.654211
624      Guinea-Bissau    Africa  1952   32.500      580653    299.850319
625      Guinea-Bissau    Africa  1957   33.489      601095    431.790457
626      Guinea-Bissau    Africa  1962   34.488      627820    522.034373
627      Guinea-Bissau    Africa  1967   35.492      601287    715.580640
628      Guinea-Bissau    Africa  1972   36.486      625361    820.224588
629      Guinea-Bissau    Africa  1977   37.465      745228    764.725963
630      Guinea-Bissau    Africa  1982   39.327      825987    838.123967
631      Guinea-Bissau    Africa  1987   41.245      927524    736.415392
632      Guinea-Bissau    Africa  1992   43.266     1050938    745.539871
633      Guinea-Bissau    Africa  1997   44.873     1193708    796.664468
634      Guinea-Bissau    Africa  2002   45.504     1332459    575.704718
635      Guinea-Bissau    Africa  2007   46.388     1472041    579.231743
984             Mexico  Americas  1952   50.789    30144317   3478.125529
985             Mexico  Americas  1957   55.190    35015548   4131.546641
986             Mexico  Americas  1962   58.299    41121485   4581.609385
987             Mexico  Americas  1967   60.110    47995559   5754.733883
988             Mexico  Americas  1972   62.361    55984294   6809.406690
989             Mexico  Americas  1977   65.032    63759976   7674.929108
990             Mexico  Americas  1982   67.405    71640904   9611.147541
991             Mexico  Americas  1987   69.498    80122492   8688.156003
992             Mexico  Americas  1992   71.455    88111030   9472.384295
993             Mexico  Americas  1997   73.670    95895146   9767.297530
994             Mexico  Americas  2002   74.902   102479927  10742.440530
995             Mexico  Americas  2007   76.195   108700891  11977.574960

## select() para crear subconjuntos de datos en variables o columnas.

Utiliza ``select()`` para crear un subconjunto de los datos de las variables.

**Objetivos:**

* Selecciona las variables ``lifeExp`` y ``year``. Nota que el orden en que estas columnas aparecen en ``gapminder`` es primero ``year`` y luego ``lifeExp``
* Obtener las variables ``year`` y ``lifeEx`` para Camboya (Camboya en inglés es Cambodia)

In [34]:
# Segmeno solo 2 columnas, pero me respeta el orden en que yo solicito
# ver los datos

gapminder >> select(_.lifeExp, _.year)

lifeExp  year
0      28.801  1952
1      30.332  1957
2      31.997  1962
3      34.020  1967
4      36.088  1972
...       ...   ...
1699   62.351  1987
1700   60.377  1992
1701   46.809  1997
1702   39.989  2002
1703   43.487  2007

[1704 rows x 2 columns]

In [40]:
gapminder >> filter(_.country == "Cambodia") >> select(_.year,_.lifeExp, _.country) 

# gapminder[gapminder["country"] == Cambodia][["year","lifeExp","country"]]

year  lifeExp   country
216  1952   39.417  Cambodia
217  1957   41.366  Cambodia
218  1962   43.415  Cambodia
219  1967   45.415  Cambodia
220  1972   40.317  Cambodia
221  1977   31.220  Cambodia
222  1982   50.957  Cambodia
223  1987   53.914  Cambodia
224  1992   55.803  Cambodia
225  1997   56.534  Cambodia
226  2002   56.752  Cambodia
227  2007   59.723  Cambodia

Nota que el orden de los verbos altera el resultado: si primero seleccionamos ``year`` y ``lifeExp``, el dataframe resultante no tendrá columna ``country``,  de modo que no podremos recuperar Camboya

In [38]:
gapminder >> select(_.year,_.lifeExp) >> filter(_.country == "Cambodia") 

AttributeError: 'DataFrame' object has no attribute 'country'

### Los métodos avanzados de selección dentro de pandas

Considere el siguiente dataframe

In [41]:
np.random.seed(2021)
mi_dataframe = {"nombre":["Luis","María","Helena"],
                "Sexo":["masc","feme","feme"],
                "Materia_matemáticas":np.random.randint(0,11,3),
                "Materia_español":np.random.randint(0,11,3),
                "Materia_inglés":np.random.randint(0,11,3),
                "Materia_computación":np.random.randint(0,11,3),
                "Evaluación_trimestral":np.random.randint(0,11,3),
                "Evaluación_general":np.random.randint(0,11,3),
                "Trimestral_aprovechamiento":["No","R","No"],
                "General_aprovechamiento":["No","No","Sí"]}

mi_dataframe = pd.DataFrame(mi_dataframe)
mi_dataframe

nombre  Sexo  Materia_matemáticas  Materia_español  Materia_inglés  \
0    Luis  masc                    4                0               8   
1   María  feme                    5                6               6   
2  Helena  feme                    9                5               6   

   Materia_computación  Evaluación_trimestral  Evaluación_general  \
0                    6                      5                   1   
1                    6                      7                   5   
2                    1                      1                  10   

  Trimestral_aprovechamiento General_aprovechamiento  
0                         No                      No  
1                          R                      No  
2                         No                      Sí

Hay ocasiones en las cuales debemos seleccionar muchas columnas de un dataframe. Pueden ser tantas, que escribir una por una dentro de un select puede llegar a ser tedioso.

**Objetivos:**

* Mostrar únicamente el nombre y las calificaciones de las materias de las mujeres.
* Mostrar únicamente el nombre y las calificaciones de las materias y las evaluaciones de las mujeres.
* Mostrar únicamente el nombre, las evaluaciones y los aprovechamientos.

In [ ]:
mi_dataframe >> filter(_.Sexo == "feme") >> select(_.nombre, _.startswith("Materia"))

In [ ]:
mi_dataframe >> filter(_.Sexo == "feme") >> select(_.nombre, _.startswith(("Evaluación","Materia")))

In [ ]:
mi_dataframe >> select(_.nombre, _.startswith("Evaluación"),_.endswith("aprovechamiento"))

--------
--------
**Tarea.**

Intenta crear la tabla anterior utilizando únicamente ``pandas`` y ``startswith``.

--------
--------

En [esta página](https://siuba.readthedocs.io/en/latest/api_table_core/03_select.html) encontrarás varios métodos de selección en siuba.

## mutate() para agregar nuevas variables

Imagina que quisiéramos recuperar el PIB de cada país. Después de todo, los datos de ``gapminder`` tienen una variable para población (``pop``) y PIB per cápita (``gdpPercap``). 

``mutate()`` es una función que define e inserta nuevas variables en un dataframe. Puedes hacer referencia a las variables existentes por su nombre.

In [ ]:
gapminder >> mutate(gdp = _.gdpPercap * _.population)

In [ ]:
## Quitando la notación científica en todos los dataframes de la sesión actual
pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [ ]:
gapminder >> mutate(gdp = _.gdpPercap * _.population)

In [ ]:
## Eliminando el formato de dos decimales que afecta globalmente
pd.reset_option('display.float_format')

In [ ]:
## Quitando la notación científica únicamente para la columna de interés

gapminder >> mutate(gdp = (_.gdpPercap * _.population).apply(lambda x: '%.2f' % x))

Esas cifras del PIB son casi inútilmente grandes y abstractas: "Una cosa que me molesta son los números grandes que se presentan sin contexto… Si añado un cero a este número, ¿la oración que lo contiene significaría algo diferente para mí? Si la respuesta es 'no', tal vez el número no tiene por qué estar en la oración en primer lugar".

Tal vez sería más significativo para los consumidores de tus tablas y cifras ceñirse al PIB per cápita. Pero, ¿y si reportara el PIB per cápita *relativo con algún país de referencia*? Ya que México es mi hogar, me quedo con eso. Es decir, cuánto representa el PIB per cápita de cada país respecto del de México. 

Necesito crear una nueva variable que se gdpPercapdivida por el de México, cuidando que siempre divida dos números que pertenezcan al mismo año.

In [ ]:
(gapminder >> filter(_.country == "Mexico"))["gdpPercap"].tolist()

In [ ]:
gdpPercap_Mex = (gapminder >> filter(_.country == "Mexico"))["gdpPercap"].tolist()

In [ ]:
gapminder.shape[0]/12

In [ ]:
total_paises = int(gapminder.shape[0]/12)

In [ ]:
total_paises*gdpPercap_Mex

In [ ]:
mi_gapminder = (gapminder >> 
    mutate(gdpPercap_respect_Mex = _.gdpPercap*100/(total_paises*gdpPercap_Mex))
    )

**Objetivos:**

* Mostrar los renglones de los países que estuvieron mejor que México
* Mostrar los renglones de los países americanes que estuvieron mejor que México
* Mostrar cómo evolucionó la economía de Venezuela respecto de la de México 

In [ ]:
mi_gapminder >> filter(_.gdpPercap_respect_Mex > 100)

In [ ]:
mi_gapminder >> filter(_.gdpPercap_respect_Mex > 100, _.continent == "Americas")

In [ ]:
mi_gapminder >> filter(_.country == "Venezuela")

## arrange() para ordenar datos por filas de una manera basada en principios

``arrange()`` reordena las filas en un dataframe.

**Objetivo:**

* Ordenar los datos por año y luego por país, en lugar de por país y luego por año.
* ¿O tal vez solo desea los datos de 2007, ordenados según la esperanza de vida?
* Ah, ¿te gustaría clasificar la esperanza de vida en orden descendente ?

In [ ]:
gapminder >> arrange(_.year,_.country)

In [ ]:
gapminder >> filter(_.year == 2007) >> arrange(_.lifeExp)

In [ ]:
gapminder >> arrange(-_.lifeExp)

Te aconsejo que tus análisis NUNCA se basen en filas o variables en un orden específico. Pero una vez que estés preparando tablas para ojos humanos, es imperativo que tomes el control del orden de las filas.

## rename() para renombrar variables

¡Cambiemos el nombre de algunas variables!

**Objetivo:**

*  Renombrar a las columnas ``country`` y ``gdpPercap`` por su nombre en español

In [ ]:
gapminder >> rename(pais = _.country, pib_percap = _.gdpPercap)

### select() también puede utilizarse para cambiar los nombres de las columnas

Has visto el uso simple de ``select()``. Hay otro truco que te puede gustar:

* ``select()`` puede cambiar el nombre de las variables que solicita mantener. Puedes renombrar una columna en específico utilizando el operador ``==`` de la siguiente manera:

``_.nuevo_nombre == _.viejo_nombre``

In [ ]:
(gapminder >>
    filter(_.country == "Burundi", _.year > 1996) >>
    select(_.yr == _.year, _.lifeExp, _.gdpPercap)
    ) 

## group_by() es un arma poderosa

Todos hemos tenido colaboradores a los cuales les encanta hacer preguntas aparentemente simples como "¿qué país experimentó la mayor caída en la esperanza de vida en 5 años?". De hecho, esa es una pregunta totalmente natural. Pero si no estás utilizando un lenguaje que procese datos, es una pregunta increíblemente molesta de responder.

**siuba** ofrece herramientas poderosas para resolver este tipo de problema.

* ``group_by()`` agrega estructura adicional a su conjunto de datos (información de agrupación) que sienta las bases para los cálculos dentro de los grupos.

* ``summarize()`` toma un conjunto de datos con $n$ observaciones, calcula los resúmenes solicitados y devuelve un conjunto de datos con 1 observación.

* ``mutate()`` y ``summarize()`` distinguen entre grupos.

### Contando cosas

Comencemos con el conteo simple. ¿Cuántas observaciones tenemos por continente?

In [ ]:
gapminder >> group_by(_.continent) >> count()

Alternativamente, podemos utilizar varias funciones incluídas en ``siuba.dply.vector``. Puedes ver muchas de ellas en https://siuba.readthedocs.io/en/latest/api_extra/vector.html

En particular, su función de conteo es ``n()``

In [ ]:
gapminder >> group_by(_.continent) >> summarize(total = n(_))

¿Qué pasaría si quisiéramos contar la cantidad de países únicos para cada continente? Puedes calcular múltiples resúmenes dentro de summarize(). 

Utiliza  ``n_distinct()`` para contar el número de países distintos dentro de cada continente.

In [ ]:
gapminder >> group_by(_.continent) >> summarize(total = n(_),paises_distintos = n_distinct(_.country))

## Resúmenes más generales

Las funciones que aplicará ``summarize()`` incluyen resúmenes estadísticos clásicos. En particular, cualquiera de las que encontramos en la sección de funciones estadísticas de **numpy**

In [ ]:
gapminder >> group_by(_.continent) >> summarize(mean_lifeExp = _.lifeExp.mean())

In [ ]:
mi_gap = gapminder.copy()

In [ ]:
mi_gap["lifeExp"][0] = np.nan

In [ ]:
mi_gap >> group_by(_.continent) >> summarize(mean_lifeExp = _.lifeExp.mean())

Centrémonos sólo en Asia. ¿Cuáles son las esperanzas de vida mínima y máxima vistas por año?

In [ ]:
(gapminder >> 
    filter(_.continent == "Asia") >> 
    group_by(_.year) >> 
    summarize(esp_min = _.lifeExp.min(), esp_max=_.lifeExp.max())
    )

Por supuesto, sería mucho más interesante ver qué país contribuyó con estas observaciones extremas. ¿El mínimo (máximo) viene siempre del mismo país? Abordaremos eso en breve.

## Mutación agrupada

A veces no deseas colapsar  las $n$ filas de cada grupo en una sola fila, sino mantener tus grupos, pero calculando nuevas columnas dentro de ellos.

Hagamos una nueva variable que sean los años de esperanza de vida ganados (perdidos) en relación con 1952, para cada país individual. Por ejemplo, para el caso de México, queremos la siguiente tabla:

country|continent|year|lifeExp|population|gdpPercap|gain_lifeExp
:-:|:-:|:-:|:-:|:-:|:-:|:-:|
Mexico|Americas|1952|50.789|30144317|3478.125529|0
Mexico|Americas|1957|55.19|35015548|4131.546641|4.401
Mexico|Americas|1962|58.299|41121485|4581.609385|7.51
Mexico|Americas|1967|60.11|47995559|5754.733883|9.321
Mexico|Americas|1972|62.361|55984294|6809.40669|11.572
Mexico|Americas|1977|65.032|63759976|7674.929108|14.243
Mexico|Americas|1982|67.405|71640904|9611.147541|16.616
Mexico|Americas|1987|69.498|80122492|8688.156003|18.709
Mexico|Americas|1992|71.455|88111030|9472.384295|20.666
Mexico|Americas|1997|73.67|95895146|9767.29753|22.881
Mexico|Americas|2002|74.902|102479927|10742.44053|24.113
Mexico|Americas|2007|76.195|108700891|11977.57496|25.406


Agrupamos por país y usamos ``mutate()`` para hacer una nueva variable. La función ``first()`` extrae el primer valor de un vector. 

Observa que ``first()`` está operando sobre el vector de esperanza de vida dentro de cada grupo de países.

In [ ]:
gapminder >> group_by(_.country) >> mutate(gain_lifeExp = _.lifeExp-first(_.lifeExp))

Si ya no harás ninguna operación extra en el dataframe anterior que involucre grupos, conviene cerrar el agrupamiento con el verbo ``ungroup()``

In [ ]:
gapminder >> group_by(_.country) >> mutate(gain_lifeExp = _.lifeExp-first(_.lifeExp)) >> ungroup()

Por el contrario, si todavía vas a trabajar con los grupos, recuerda que tanto ``mutate()`` como ``summarize()`` distinguen los grupos.

In [ ]:
(gapminder >>
    group_by(_.country) >> 
    mutate(gain_lifeExp = _.lifeExp-first(_.lifeExp)) >>
    summarize(mean_gain_lifeExp = _.gain_lifeExp[1:].mean())
    )